### DISCUSSION TITLE
How to modify the porosity, permeability and wettability of the model in OpenPNM?

### DISCUSSION THREAD

**Discussion by Xiaozizhuo** [PMEAL/OpenPNM#2357] [Q&A] [answered]
I reproduced many examples in OpenPNM and wanted to extend on the basis of the examples, but so far only the pore diameter and throat length and diameter have been modified. How to modify its porosity, wettability, permeability and other physical parameters of porous media?  

If we take this as an example, how should we modify it?
![3](https://user-images.githubusercontent.com/98071787/160237695-54aa901c-21ac-48cb-8087-45583602d845.png)

**Comment by ma-sadeghi** [comment, marked as answer]
Well, by changing pore/throat diameter, you're essentially changing the porosity, and subsequently the permeability. You can't directly change the permeability of a pore network model (or any other pore-scale method), that's what you do in a volume-averaged continuum model. As for wettability, I'm not very familiar, @jgostick knows more. But, I guess it has to do with the phase properties (ex. you can modify contact angle in your phase, say `op.phase.Mercury`).

**Reply by Xiaozizhuo** [reply]
Ah yes, I almost forgot the contact angle, an important physical quantity of porous media. So how to modify the contact angle in the above code? It seems that my code does not mention information related to contact angle? By the way, how to make the shape of pores and throats not limited to spheres and cylinders?

**Reply by ma-sadeghi** [reply]
Well, you must first add a phase. We store contact angle information on phase objects.

```python
air = op.phase.Air(network=pn)
air.remove_model("pore.contact_angle")
air["pore.contact_angle"] = 50.0
```

Note 1: the reason you have to remove the existing model first is that if you don't, the model might get regenerated and the value you manually set gets overridden. 

Note 2: If you want to see a change in the simulation outcome, I think you need to set a distribution of contact angles, otherwise if it's all uniform, it won't affect the outcome (I'm not sure about this though, double check).

**Reply by Xiaozizhuo** [reply]
Thanks for your help, the code runs successfully. If I want to calculate what the porosity of a model is, how should I do it?

**Reply by ma-sadeghi** [reply]
My pleasure. Calculating porosity is very tricky because of the following reasons:

- Each pair of pores overlap with the connecting throat
- Adjacent pores might overlap in some scenarios where throat length ~ 0
- Throats can also overlap for highly connected networks (ex. diagonally connected throats in a cubic lattice can easily overlap)
- And more importantly, the total volume is vaguely defined. You might say it's the bounding box, but where that box begins is up for debate, plus for unstructured networks, it's really difficult to define what we mean by the bounding box.

In principle, it seems that these overlaps can be calculated and subtracted from the total volume, but so far no one has done it!

That being said, here's a naive way to calculate porosity in Cubic networks, at least it gives an upper bound (but it could be way off in some scenarios):

```
import numpy as np
import openpnm as op

volume_void =  net["pore.volume"].sum() + net["throat.volume"].sum()
shape = op.topotools.get_shape(net)
spacing = op.topotools.get_spacing(net)
volume_total = np.prod(shape * spacing)

porosity = volume_void / volume_total
```

**Reply by Xiaozizhuo** [reply]
It is a great pleasure to meet a researcher like you who is eager to help you during the preliminary research on numerical simulation of porous media.
I ran your code and got the following error, is it a version problem? My version is 2.8.2, and I seem to have similar errors in the code on many examples.
![error1](https://user-images.githubusercontent.com/98071787/160282475-b0b506ee-1166-4065-889e-d00f78229310.png)

**Reply by ma-sadeghi** [reply]
I'm happy to help. We're transitioning to v3 soon. For now, please install the development version. See [here](https://openpnm.org/user_guide/installation.html#installing-the-dev-version) for instructions.

**Reply by Xiaozizhuo** [reply]
I installed OpenPNM according to the example on your official website, but still the same error is reported.
   My graduation thesis this time is to use OpenPNM to make a GDL, but now I find that there are a lot of problems that cannot be run due to version problems. Not sure if porespy can do this
   Nonetheless, your kind help has given me a lot of encouragement
![error1](https://user-images.githubusercontent.com/98071787/160343887-29aca746-288f-46fe-a1e6-1ba0f7f37c11.png)
![error2](https://user-images.githubusercontent.com/98071787/160343915-96f01090-5e35-4872-9e47-cebf8b8a8792.png)

**Reply by ma-sadeghi** [reply]
Just to make sure, please do so one more time this time in a virtual environment (I assume you already have Anaconda distribution installed):

```bash
conda create -n testenv python=3.8 mamba -c conda-forge --override-channels
conda activate testenv
git clone https://github.com/PMEAL/OpenPNM
mamba install --file OpenPNM/requirements/conda_requirements.txt -c conda-forge --override-channels
pip install --no-deps -e OpenPNM/
```

I did this in a fresh virtual env and it worked

**Reply by Xiaozizhuo** [reply]
This method seems to be cumbersome for my computer, causing various errors. I wonder if it is possible to update the `topotools` function file in my older version by sending the file of the `topotools` function in the development version of OpenPNM. Because my topotools does not define what is `get_shape` and `get_spacing`

**Reply by ma-sadeghi** [reply]
In principle, you can. Just copy those two functions and paste it in your own topotools. But, it might need some tweaking.

Let's try one last time:
1. Delete your current `openpnm`: Run `pip uninstall openpnm` and `conda remove openpnm`
2. Install the development version the easy way: `pip install git+https://github.com/PMEAL/OpenPNM.git@dev`

If this also doesn't work, I highly recommend that you remove your Anaconda distribution and reinstall it, as there might be some serious conflicts already in your base environment.

**Reply by Xiaozizhuo** [reply]
Sorry, I still failed. Is there any other way to calculate porosity? It can be rough, such as only calculating the ratio of pore volume to total volume

**Reply by ma-sadeghi** [reply]
```python
import numpy as np
import openpnm as op

shape = [5, 6, 7]
spacing = 1e-5

net = op.network.Cubic(shape, spacing)
geo = op.geometry.SpheresAndCylinders(network=net, pores=net.Ps, throats=net.Ts)

volume_void =  net["pore.volume"].sum() + net["throat.volume"].sum()
volume_total = np.prod(shape) * spacing**3

porosity = volume_void / volume_total
print(f"Porosity: {porosity:.3f})
```

**Reply by Xiaozizhuo** [reply]
Sorry, before seeing your reply. I found the definitions of these two functions on the official website of OpenPNM, so I downloaded all the files on the official website and replaced the previous files. As a result, the following problem now occurs.
![error1](https://user-images.githubusercontent.com/98071787/160633343-e8251f86-503a-4a58-bcfd-aa2b78e87e4e.png)
![error2](https://user-images.githubusercontent.com/98071787/160633371-12dda1ea-f7e6-47a6-9701-00126d43bea5.png)
I'm really sorry I've been bothering you

**Reply by ma-sadeghi** [reply]
It's really hard to tell. The only clean option that comes to mind is:

1. Delete your Python installation (Anaconda, if that's your distribution)
2. Remove all folders related to OpenPNM
3. Install the [Anaconda](https://www.anaconda.com/products/distribution) Python distribution
4. Open Anaconda Command Prompt and run this command:
5. `pip install git+https://github.com/PMEAL/OpenPNM.git@dev`
